src/test/ui/stability-attribute/generics-default-stability.rs

```diff
@@ -0,0 +1,106 @@
+// ignore-tidy-linelength
+// aux-build:unstable_generic_param.rs
+
+extern crate unstable_generic_param;
+
+use unstable_generic_param::*;
+
+struct R;
+
+impl Trait1 for S {
+    fn foo() -> () { () } // ok
+}
+
+struct S;
+
+impl Trait1<usize> for S { //~ ERROR use of unstable library feature 'unstable_default'
+    fn foo() -> usize { 0 }
+}
+
+impl Trait1<isize> for S { //~ ERROR use of unstable library feature 'unstable_default'
+    fn foo() -> isize { 0 }
+}
+
+impl Trait2<usize> for S { //~ ERROR use of unstable library feature 'unstable_default'
+    fn foo() -> usize { 0 }
+}
+
+impl Trait3<usize> for S {
+    fn foo() -> usize { 0 } // ok
+}
+
+fn main() {
+    let _ = S;
+
+    let _: Struct1<isize> = Struct1 { field: 1 }; //~ ERROR use of unstable library feature 'unstable_default'
+
+    let _ = STRUCT1; // ok
+    let _: Struct1 = STRUCT1; // ok
+    let _: Struct1<usize> = STRUCT1; //~ ERROR use of unstable library feature 'unstable_default'
+    let _: Struct1<isize> = Struct1 { field: 0 }; //~ ERROR use of unstable library feature 'unstable_default'
+
+    // Instability is not enforced for generic type parameters used in public fields.
+    // Note how the unstable type default `usize` leaks,
+    // and can be used without the 'unstable_default' feature.
+    let _ = STRUCT1.field;
+    let _ = Struct1 { field: 1 };
```

Comment:
What happens if we do:
```rust
let _ = Struct1 { field: () };
```
i.e. try to use a type that's not the default, implicitly?

Reply:
> Instability is not enforced for generic type parameters used in public fields. Note how the unstable type default `usize` leaks, and can be used without the 'unstable_default' feature.

You can change the inferred type of a public field, you just can't write it. This is https://github.com/rust-lang/rust/pull/72314#issuecomment-634285595.

Reply:
I think we want these to error. Probably you should only be able to use a non-default type if you've opted into the feature flag. I'll try to point you in the right direction for this tomorrow.

Reply:
Sounds good.
I'm still not clear why you should be able to use a unstable type parameter's default on stable?
That still feels like a leak to me.

Reply:
The goal is to hide the possibility that one of the types in a definition is actually a parameter. To the user, it should just look like a fixed type. It doesn't matter if they use the concrete type, because nothing breaks if it's technically a parameter: the user has to opt in to changing the type parameter. However, now that I think about it, if we change inference, then something could break when a type parameter default is stabilised… So maybe we need to leave it up to the API designer to make sure that the type doesn't leak.

Reply:
The most important thing is to ensure that removing the stability attribute doesn't break any code, and that removing an unstable defaulted type parameter doesn't break anyone who didn't opt into the feature gate. Maybe what you have already is the best we can do. I think observing the concrete type is okay, but changing it (without the feature flag) should not be possible. I don't see how we could forbid that, though.

Reply:
> The most important thing is to ensure that removing the stability attribute doesn't break any code, and that removing an unstable defaulted type parameter doesn't break anyone who didn't opt into the feature gate.

I added a `Box<T, A: Alloc = System>` test to that account.

> I don't see how we could forbid that, though

The options for this are "returning a dummy unstable type when the default is unstable in the `|substs, param, infer_args|` closure passed to `create_substs_for_generic_args`, or somehow marking fields unstable if they mention a unstable type parameter. I don't believe either would not be a complete fix since the type and details about the type could still leak via methods.

Reply:
Yes, I agree. I think a complete fix is looking quite implausible. It might be best to leave this to the API designers and make it very clear that they are responsible for not allowing leakage. I think it should be straightforward to avoid once they're aware of the limitations. I shall write a summary post on https://github.com/rust-lang/wg-allocators/issues/2, and then I think we're probably good to go on this PR.